src/i_winmusic.c

```diff
@@ -291,18 +316,237 @@ static void ResetDevice(void)
+            }
+            break;
+
+        case 0x7E: // Universal Non-Real Time
+            switch (msg[3])
+            {
+                case 0x09: // General Midi
+                    if (length == 6 &&
+                       (msg[4] == 0x01 || // GM System On
+                        msg[4] == 0x02 || // GM System Off
+                        msg[4] == 0x03))  // GM2 System On
+                    {
+                        // GM System On/Off, GM2 System On
+                        // F0 7E <dev> 09 01 F7
+                        // F0 7E <dev> 09 02 F7
+                        // F0 7E <dev> 09 03 F7
+                        return true;
+                    }
+                    break;
+            }
+            break;
+    }
+    return false;
+}
+
+static boolean IsMasterVolume(byte *msg, int length, unsigned int *volume)
+{
+    // General Midi (F0 7F <dev> 04 01 <lsb> <msb> F7)
+    if (length == 8 &&
+        msg[1] == 0x7F && // Universal Real Time
+        msg[3] == 0x04 && // Device Control
+        msg[4] == 0x01)   // Master Volume
+    {
+        *volume = (msg[5] & 0x7F) | ((msg[6] & 0x7F) << 7);
+        return true;
+    }
+
+    // Roland (F0 41 <dev> 42 12 40 00 04 <vol> <sum> F7)
+    if (length == 11 &&
+        msg[1] == 0x41 && // Roland
+        msg[3] == 0x42 && // GS
+        msg[4] == 0x12 && // DT1
+        msg[5] == 0x40 && // Address MSB
+        msg[6] == 0x00 && // Address
+        msg[7] == 0x04)   // Address LSB
+    {
+        *volume = DEFAULT_MASTER_VOLUME * (msg[8] & 0x7F) / 127;
+        return true;
+    }
+
+    // Yamaha (F0 43 <dev> 4C 00 00 04 <vol> F7)
+    if (length == 9 &&
+        msg[1] == 0x43 && // Yamaha
+        msg[3] == 0x4C && // XG
+        msg[4] == 0x00 && // Address High
+        msg[5] == 0x00 && // Address Mid
+        msg[6] == 0x04)   // Address Low
+    {
+        *volume = DEFAULT_MASTER_VOLUME * (msg[7] & 0x7F) / 127;
+        return true;
+    }
+
+    return false;
+}
+
+static boolean SendSysExMsg(int time, midi_event_t *event)
+{
+    // SysEx messages in MIDI files (Standard MIDI Files 1.0 pages 6-7):
+    // Complete:        (F0 ... F7)
+    // Multi-packet:    (F0 ...) + (F7 ...) + ... + (F7 ... F7)
+    // Escape sequence: (F7 ...)
+
+    if (event->data.sysex.length + sysex_buffer_length > MAX_SYSEX_LENGTH - 1)
+    {
+        // Ignore messages that are too long.
+        sysex_buffer_length = 0;
+        return false;
+    }
+
+    if (event->event_type == MIDI_EVENT_SYSEX_SPLIT && sysex_buffer_length == 0)
+    {
+        // Ignore escape sequences (don't send illegal characters).
+        return false;
+    }
+
+    if (event->event_type == MIDI_EVENT_SYSEX)
+    {
+        // Start a new message (discards any previous incomplete message).
+        sysex_buffer[0] = MIDI_EVENT_SYSEX;
+        sysex_buffer_length = 1;
+    }
+
+    // Copy message to tracked buffer.
+    memcpy(&sysex_buffer[sysex_buffer_length], event->data.sysex.data,
```

Comment:
I think there is no need to copy the SysEx message here, you can use the `event->data.sysex.data` buffer.

Reply:
> My understanding is that master volume SysEx is part of the GM standard and should be supported by everyone, right? At least all the softsynths I've tried work.

Yes, every device should be able to receive master volume message unless the user has gone out of their way to disable the feature.

Reply:
From the comment:
```
    // SysEx messages in MIDI files (Standard MIDI Files 1.0 pages 6-7):
    // Complete:        (F0 ... F7)
    // Multi-packet:    (F0 ...) + (F7 ...) + ... + (F7 ... F7)
    // Escape sequence: (F7 ...)
```
Copying to `sysex_buffer` and tracking the length is in case there's a SysEx reset or master volume split across multiple packets. If I only checked a single packet I could miss the message and the volume would be reset.

Reply:
Oh, so a master volume message can spread across multiple packets? Okay, now I understand.

Reply:
I doubt someone would do that, but just to be safe.

Reply:
Yes, we may be overcomplicating this. Does the mater volume message spread across multiple packages even work on synth/hardware? I feel like it will be dropped.

Reply:
I tried this:
```diff
diff --git a/src/i_winmusic.c b/src/i_winmusic.c
index 6d3ad80..703f9e1 100644
--- a/src/i_winmusic.c
+++ b/src/i_winmusic.c
@@ -52,8 +52,12 @@ static byte xg_system_on[] = {
     0xF0, 0x43, 0x10, 0x4C, 0x00, 0x00, 0x7E, 0x00, 0xF7
 };
 
-static byte master_volume_msg[] = {
-    0xF0, 0x7F, 0x7F, 0x04, 0x01, 0x7F, 0x7F, 0xF7
+static byte master_volume_msg1[] = {
+    0xF0, 0x7F, 0x7F, 0x04, 0x01
+};
+
+static byte master_volume_msg2[] = {
+    0xF7, 0x7F, 0x7F, 0xF7
 };
 
 #define DEFAULT_MASTER_VOLUME 16383
@@ -258,9 +262,10 @@ static void UpdateTempo(int time)
 static void UpdateVolume(int time)
 {
     int volume = master_volume * volume_factor + 0.5f;
-    master_volume_msg[5] = volume & 0x7F;
-    master_volume_msg[6] = (volume >> 7) & 0x7F;
-    SendLongMsg(time, master_volume_msg, sizeof(master_volume_msg));
+    master_volume_msg2[1] = volume & 0x7F;
+    master_volume_msg2[2] = (volume >> 7) & 0x7F;
+    SendLongMsg(time, master_volume_msg1, sizeof(master_volume_msg1));
+    SendLongMsg(time, master_volume_msg2, sizeof(master_volume_msg2));
 }
 
 static void ResetDevice(void)
```
It doesn't work.

Reply:
I can't get it to work either. It looks like only some Casio equipment ever did this. How about we ignore any SysEx message starting with 0xF7 and also ignore any message that doesn't end in 0xF7?

Reply:
> How about we ignore any SysEx message starting with 0xF7 and also ignore any message that doesn't end in 0xF7?

Yes, ok. Or don't check at all, it's very easy to make a WAD that crashes Doom 😉

Reply:
Fixed by https://github.com/fabiangreffrath/woof/pull/822/commits/78b91c597c2ec78031e586c9ff7ec4c51c20ed5b